Thanks for addressing the previous report in `226d023`. The original `RDF('0.499999994')` example is now positive. However, the same underlying issue remains for exact rational or higher-precision Sage inputs because `delta` is converted to `RDF` before validation and before the distance from `1/2` is formed.

### Exact inputs just below `1/2` still return zero

For example:

```python
sage: delta = 1/2 - 2^-100
sage: delta < 1/2
True
sage: RDF(delta)
0.5
sage: codes.bounds.mrrw2_bound_asymp(delta, 2)
0.0
```

Here

$$
1-2\delta=2^{-99},
$$

and high-precision evaluation of the right endpoint gives

$$
g((1-2\delta)^2)
\approx 1.253580968852975 \times 10^{-58}>0.
$$

This value is well within the representable range of `RDF`. Thus, returning zero is not an unavoidable limitation of double precision; the relevant information is lost specifically when `delta` is rounded to `0.5` before `upper = 1 - 2*delta` is computed. In this distance range, the original paper states that the right endpoint supplies the minimum.

The same early conversion also bypasses the documented range check:

```python
sage: delta = 1/2 + 2^-100
sage: delta > 1/2
True
sage: codes.bounds.mrrw2_bound_asymp(delta, 2)
0.0
```

This should raise `ValueError`.

Could the range check and exact endpoint checks be performed before converting to `RDF`, and could `upper` be formed from the original input precision? For example, the implementation could preserve

```python
upper = RDF(1 - 2*delta)
```

before any lossy conversion. The second argument of `g` in the objective can also be written as

```python
b = 1 - (1 + u) * (upper - u)
```

since this is algebraically equal to

```python
u**2 + 2*delta*u + 2*delta
```

and avoids reconstructing the expression from a `delta` that may already have rounded to `0.5`.

Suggested regression tests are:

```python
sage: codes.bounds.mrrw2_bound_asymp(1/2 - 2^-100, 2) > 0  # needs scipy
True
sage: codes.bounds.mrrw2_bound_asymp(1/2 + 2^-100, 2)
Traceback (most recent call last):
...
ValueError: the relative minimum distance must be between 0 and 1/2
```

### Remaining relative-accuracy issue in `g`

There is also a smaller numerical-accuracy issue. For

```python
RDF('0.499999994')
```

the current implementation returns

```text
1.9664921034196128e-15
```

whereas high-precision evaluation at the same binary floating-point value gives approximately

```text
2.018429124836941e-15
```

a relative error of about 2.6%. The transformed entropy argument avoids the first cancellation, but the existing `entropy` implementation subsequently forms `1-p`; when `p` is sufficiently small, this rounds to `1.0` and drops the corresponding entropy term. A binary-entropy evaluation using `log1p(-p)` would avoid that loss. If relative accuracy near `1/2` is intended, it would also be better for the regression test to check the expected numerical value rather than only `> 0`.

Apart from these endpoint numerical issues, the formula, public export, reference, ordinary sample values, and endpoint comparison look good.

--- a/src/sage/coding/code_bounds.py
+++ b/src/sage/coding/code_bounds.py
@@ -163,7 +163,7 @@ PROBLEM: In this module we shall typically either (a) seek bounds on `k`, given
 #                  https://www.gnu.org/licenses/
 # ****************************************************************************
 
-from math import log1p
+from math import exp, log1p
 
 from sage.misc.lazy_import import lazy_import
 from sage.rings.real_double import RDF
@@ -797,15 +797,23 @@ def mrrw2_bound_asymp(delta, q):
 
         sage: codes.bounds.mrrw2_bound_asymp(0, 2)
         1.0
+        sage: codes.bounds.mrrw2_bound_asymp(2^-60, 2)
+        1.0
         sage: codes.bounds.mrrw2_bound_asymp(1/2, 2)
         0.0
 
     Values below the upper endpoint retain their numerical accuracy::
 
-        sage: codes.bounds.mrrw2_bound_asymp(RDF('0.499999994'), 2)  # abs tol 1e-27    # needs scipy
+        sage: # needs scipy
+        sage: codes.bounds.mrrw2_bound_asymp(RDF('0.499999994'), 2)  # abs tol 1e-27
         2.018429124836941e-15
-        sage: codes.bounds.mrrw2_bound_asymp(1/2 - 2^-100, 2)  # abs tol 1e-70          # needs scipy
+        sage: codes.bounds.mrrw2_bound_asymp(1/2 - 2^-100, 2)  # abs tol 1e-70
         1.2535809688529749e-58
+        sage: codes.bounds.mrrw2_bound_asymp(1/2 - 2^-538, 2) > 0
+        True
+        sage: all(codes.bounds.mrrw2_bound_asymp(1/2 - 2^-n, 2) > 0
+        ....:     for n in range(539, 543))
+        True
 
     The second bound is only known for binary codes::
 
@@ -841,17 +849,31 @@ def mrrw2_bound_asymp(delta, q):
     if twice_delta == 1:
         return RDF(0)
     upper = RDF(1 - twice_delta)
+    if upper == 1:
+        return RDF(1)
     log_two = RDF.log2()
+    smallest_normal = RDF(2)**-1022
 
     def binary_entropy(p):
         if p == 0:
             return RDF(0)
         return RDF((-p*log(p) + (p - 1)*log1p(-p)) / log_two)
 
-    def g(x):
+    def binary_entropy_from_log_p(log_p):
+        # H_2(p) = p*(1 - log(p))/log(2) + O(p^2), and the error is below
+        # the range of RDF here.
+        log_entropy = log_p + log(1 - log_p) - log(log_two)
+        return RDF(exp(log_entropy))
+
+    def g(x, log_x=None):
         x = RDF(x)
         s = sqrt(1 - x)
-        return binary_entropy(x / (2 * (1 + s)))
+        p = x / (2 * (1 + s))
+        if p < smallest_normal and (x != 0 or log_x is not None):
+            if log_x is None:
+                log_x = log(x)
+            return binary_entropy_from_log_p(log_x - 2*log_two)
+        return binary_entropy(p)
 
     def one_minus_g(y):
         y = RDF(y)
@@ -876,5 +898,5 @@ def mrrw2_bound_asymp(delta, q):
                              options={'xatol': 1e-12})
     if not result.success:
         raise RuntimeError("unable to minimize the second MRRW bound")
-    upper_value = g(upper**2)
+    upper_value = g(upper**2, 2*log(upper))
     return RDF(min(objective(0), result.fun, upper_value))

--- a/src/sage/coding/code_bounds.py
+++ b/src/sage/coding/code_bounds.py
@@ -163,16 +163,21 @@ PROBLEM: In this module we shall typically either (a) seek bounds on `k`, given
 #                  https://www.gnu.org/licenses/
 # ****************************************************************************
 
-from sage.arith.misc import binomial, is_prime_power
-from sage.features.gap import GapPackage
-from sage.misc.functional import sqrt, log
+from math import log1p
+
 from sage.misc.lazy_import import lazy_import
-from sage.rings.integer_ring import ZZ
-from sage.rings.rational_field import QQ
 from sage.rings.real_double import RDF
 
-from .delsarte_bounds import (delsarte_bound_hamming_space,
-                              delsarte_bound_additive_hamming_space)
+from sage.arith.misc import binomial, is_prime_power
+from sage.features.gap import GapPackage
+from sage.misc.functional import log, sqrt
+from sage.rings.integer_ring import ZZ
+from sage.rings.rational_field import QQ
+
+from .delsarte_bounds import (
+    delsarte_bound_additive_hamming_space,
+    delsarte_bound_hamming_space,
+)
 
 lazy_import('sage.libs.gap.libgap', 'libgap')
 
@@ -795,10 +800,12 @@ def mrrw2_bound_asymp(delta, q):
         sage: codes.bounds.mrrw2_bound_asymp(1/2, 2)
         0.0
 
-    Values below the upper endpoint remain positive::
+    Values below the upper endpoint retain their numerical accuracy::
 
-        sage: codes.bounds.mrrw2_bound_asymp(RDF('0.499999994'), 2) > 0                 # needs scipy
-        True
+        sage: codes.bounds.mrrw2_bound_asymp(RDF('0.499999994'), 2)  # abs tol 1e-27    # needs scipy
+        2.018429124836941e-15
+        sage: codes.bounds.mrrw2_bound_asymp(1/2 - 2^-100, 2)  # abs tol 1e-70          # needs scipy
+        1.2535809688529749e-58
 
     The second bound is only known for binary codes::
 
@@ -817,31 +824,54 @@ def mrrw2_bound_asymp(delta, q):
         Traceback (most recent call last):
         ...
         ValueError: the relative minimum distance must be between 0 and 1/2
+        sage: codes.bounds.mrrw2_bound_asymp(1/2 + 2^-100, 2)
+        Traceback (most recent call last):
+        ...
+        ValueError: the relative minimum distance must be between 0 and 1/2
     """
     if q != 2:
         raise NotImplementedError("the second MRRW bound is only implemented "
                                   "for binary codes")
-    delta = RDF(delta)
-    if delta < 0 or delta > 0.5:
+    twice_delta = 2*delta
+    if twice_delta < 0 or twice_delta > 1:
         raise ValueError("the relative minimum distance must be between 0 "
                          "and 1/2")
-    if delta == 0:
+    if twice_delta == 0:
         return RDF(1)
-    if delta == 0.5:
+    if twice_delta == 1:
         return RDF(0)
+    upper = RDF(1 - twice_delta)
+    log_two = RDF.log2()
+
+    def binary_entropy(p):
+        if p == 0:
+            return RDF(0)
+        return RDF((-p*log(p) + (p - 1)*log1p(-p)) / log_two)
 
     def g(x):
         x = RDF(x)
         s = sqrt(1 - x)
-        return RDF(entropy(x / (2 * (1 + s)), 2))
+        return binary_entropy(x / (2 * (1 + s)))
+
+    def one_minus_g(y):
+        y = RDF(y)
+        if y == 0:
+            return RDF(0)
+        if y == 1:
+            return RDF(1)
+        t = sqrt(y)
+        if y < 0.5:
+            value = log1p(-y) + t*(log1p(t) - log1p(-t))
+        else:
+            value = (1 - t)*log1p(-t) + (1 + t)*log1p(t)
+        return RDF(value / (2*log_two))
 
     def objective(u):
         a = u**2
-        b = a + 2*delta*u + 2*delta
-        return g(a) + (1 - g(b))
+        y = (1 + u) * (upper - u)
+        return g(a) + one_minus_g(y)
 
     from scipy.optimize import minimize_scalar
-    upper = 1 - 2*delta
     result = minimize_scalar(objective, bounds=(0, upper), method='bounded',
                              options={'xatol': 1e-12})
     if not result.success:

